lib/cufft/fft.jl

```diff
@@ -382,143 +382,130 @@ end
     assert_applicable(p, X)
-    (size(Y) == p.osz) ||
+    if size(Y) != p.osz
         throw(ArgumentError("CuFFT plan applied to wrong-size output"))
-    # type errors should be impossible by dispatch, but just in case:
-    if p.xtype ∈ [CUFFT_C2R, CUFFT_Z2D]
-        (Ty == real(T)) ||
-            throw(ArgumentError("Type mismatch for argument Y"))
-    elseif p.xtype ∈ [CUFFT_R2C, CUFFT_D2Z]
-        (Ty == complex(T)) ||
-            throw(ArgumentError("Type mismatch for argument Y"))
-    else
-        (Ty == T) ||
-            throw(ArgumentError("Type mismatch for argument Y"))
+    elseif inplace != (pointer(X) == pointer(Y))
+        throw(ArgumentError(string("CuFFT ",
+                                   inplace ? "in-place" : "out-of-place",
+                                   " plan applied to ",
+                                   inplace ? "out-of-place" : "in-place",
+                                   " data")))
     end
 end
 
-function unsafe_execute!(plan::cCuFFTPlan{cufftComplex,K,true,N},
-                         x::DenseCuArray{cufftComplex,N}) where {K,N}
+function unsafe_execute!(plan::cCuFFTPlan{cufftComplex,K,<:Any,N},
+                         x::DenseCuArray{cufftComplex,N},
+                         y::DenseCuArray{cufftComplex,N}) where {K,N}
     @assert plan.xtype == CUFFT_C2C
     update_stream(plan)
-    cufftExecC2C(plan, x, x, K)
+    cufftExecC2C(plan, x, y, K)
 end
+
 function unsafe_execute!(plan::rCuFFTPlan{cufftComplex,K,true,N},
-                         x::DenseCuArray{cufftComplex,N}) where {K,N}
+                         x::DenseCuArray{cufftComplex,N},
+                         y::DenseCuArray{cufftReal,N}) where {K,N}
     @assert plan.xtype == CUFFT_C2R
     update_stream(plan)
-    cufftExecC2R(plan, x, x)
-end
-
-function unsafe_execute!(plan::cCuFFTPlan{cufftComplex,K,false,N},
-                         x::DenseCuArray{cufftComplex,N}, y::DenseCuArray{cufftComplex}
-                         ) where {K,N}
-    @assert plan.xtype == CUFFT_C2C
-    x = copy(x)  # JuliaGPU/CuArrays.jl#345, NVIDIA/cuFFT#2714055
-    update_stream(plan)
-    cufftExecC2C(plan, x, y, K)
-    unsafe_free!(x)
+    cufftExecC2R(plan, x, y)
 end
 function unsafe_execute!(plan::rCuFFTPlan{cufftComplex,K,false,N},
-                         x::DenseCuArray{cufftComplex,N}, y::DenseCuArray{cufftReal}
-                         ) where {K,N}
+                         x::DenseCuArray{cufftComplex,N},
+                         y::DenseCuArray{cufftReal}) where {K,N}
     @assert plan.xtype == CUFFT_C2R
     x = copy(x)  # JuliaGPU/CuArrays.jl#345, NVIDIA/cuFFT#2714055
     update_stream(plan)
     cufftExecC2R(plan, x, y)
     unsafe_free!(x)
 end
 
-function unsafe_execute!(plan::rCuFFTPlan{cufftReal,K,false,N},
-                         x::DenseCuArray{cufftReal,N}, y::DenseCuArray{cufftComplex,N}
-                         ) where {K,N}
+function unsafe_execute!(plan::rCuFFTPlan{cufftReal,K,<:Any,N},
```

Comment:
```suggestion
function unsafe_execute!(plan::rCuFFTPlan{cufftReal,K,<:Bool,N},
```
Would `<:Bool` here be more helpful for readers?

Reply:
Eh, I'm not sure. It puts additional constraints on dispatch where none are needed, so hypothetically it could make type intersection more complicated? I'll add the `<:Bool` to the type definition instead, for clarity.